### PR TITLE
[core] Implement image expression

### PR DESCRIPTION
**Pull request by alexshalamov**
Backport of https://github.com/mapbox/mapbox-gl-js/pull/8684

The image expression allows determining an image's availability. Comparing to a simple string, it adds semantic information, for instance, in case of formatted text, now it is possible to make distinction between image and text sections `["format", "Text", ["image", "illustration"]]`

In this PR image expression is used in places where strings were used to represent images (`*-pattern`, `icon-image`), however, in follow-up PR's image expression will be used in format expression in order to render images in labels.

https://github.com/mapbox/mapbox-gl-js/issues/8604
https://github.com/mapbox/mapbox-gl-js/pull/8904

Fixes: #15800

### CODE REVIEW

**Comment by alexshalamov** [comment]
@tobrun @julianrex @tmpsantos What is our strategy for platform bindings? I can't implement and land core part without writing bindings for all platforms :disappointed:

**Comment by alexshalamov** [comment]
@julianrex @tobrun @springmeyer Could you please take a look at [darwin], [android] and [node] commits.

File: platform/android/scripts/generate-style-code.js [hunk 1]
```diff
@@ -207,7 +207,7 @@ global.defaultValueJava = function(property) {
       case 'formatted':
         return 'new Formatted(new FormattedSection("default"))'
       case 'string':
-      case 'image': // TODO: replace once we implement image expressions
+      case 'resolvedImage': // TODO: replace once we implement image expressions
```
Comment (tobrun) [inline platform/android/scripts/generate-style-code.js]: the TODO's in this file can be removed?

Comment (alexshalamov) [reply platform/android/scripts/generate-style-code.js]: thanks, will remove.

**Comment by ryanhamley** [comment]
Thanks so much for taking this on @alexshalamov It looks great

File: include/mbgl/style/expression/image.hpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+namespace mbgl {
+namespace style {
+namespace expression {
+
+class Image {
+public:
+    Image() = default;
+    Image(const char* imageID);
+    Image(std::string imageID);
```
Comment (pozdnyakov) [inline include/mbgl/style/expression/image.hpp]: ditto

File: include/mbgl/style/expression/image.hpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    explicit Image(std::string imageID, bool available);
+    bool operator==(const Image&) const;
+    mbgl::Value serialize() const;
```
Comment (pozdnyakov) [inline include/mbgl/style/expression/image.hpp]: "serialize" is more for text or row data, maybe `toValue`?

File: include/mbgl/style/expression/type.hpp [hunk 4]
```diff
@@ -73,6 +73,11 @@ struct FormattedType {
     bool operator==(const FormattedType&) const { return true; }
 };
 
+struct ImageType {
+    constexpr ImageType(){}; // NOLINT
```
Comment (pozdnyakov) [inline include/mbgl/style/expression/type.hpp]: why is that needed^?

File: src/mbgl/renderer/image_manager.cpp [hunk 5]
```diff
@@ -169,6 +171,10 @@ void ImageManager::reduceMemoryUseIfCacheSizeExceedsLimit() {
     }
 }
 
+std::set<std::string> ImageManager::getAvailableImages() const {
```
Comment (pozdnyakov) [inline src/mbgl/renderer/image_manager.cpp]: `const std::set<std::string>& ` ?

File: src/mbgl/style/expression/is_constant.cpp [hunk 6]
```diff
@@ -49,6 +49,19 @@ bool isZoomConstant(const Expression& e) {
...
 
+bool isRuntimeConstant(const Expression& expression) {
+    if (expression.getKind() == Kind::ImageExpression) {
+        return false;
+    }
+
+    bool runtimeConstant = true;
+    expression.eachChild([&](const Expression& e) {
+        if (runtimeConstant && !isRuntimeConstant(e)) {
+            runtimeConstant = false;
```
Comment (pozdnyakov) [inline src/mbgl/style/expression/is_constant.cpp]: return false ?

File: include/mbgl/style/expression/formatted.hpp [hunk 7]
```diff
@@ -36,7 +36,9 @@ struct FormattedSection {
 class Formatted {
 public:
     Formatted() = default;
-    
+
+    Formatted(std::string str) : Formatted(str.c_str()) {}
```
[the remark this answers is not in the document]
Comment (alexshalamov) [reply include/mbgl/style/expression/formatted.hpp]: thanks, fixed.

File: include/mbgl/style/expression/expression.hpp [hunk 8]
```diff
@@ -172,8 +177,11 @@ class Expression {
 
     Kind getKind() const { return kind; };
     type::Type getType() const { return type; };
-    
-    EvaluationResult evaluate(optional<float> zoom, const Feature& feature, optional<double> colorRampParameter) const;
+
+    EvaluationResult evaluate(optional<float> zoom,
+                              const Feature& feature,
+                              optional<double> colorRampParameter,
+                              const std::set<std::string>& availableImages = {}) const;
```
[the remark this answers is not in the document]
Comment (alexshalamov) [reply include/mbgl/style/expression/expression.hpp]: done

File: include/mbgl/style/expression/image.hpp [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+#include <vector>
+
+namespace mbgl {
+namespace style {
+namespace expression {
+
+class Image {
+public:
+    Image() = default;
+    Image(const char* imageID);
```
[the remark this answers is not in the document]
Comment (alexshalamov) [reply include/mbgl/style/expression/image.hpp]: These are conversion constructors. Image class is a tiny wrapper on top of string, much easier to work with generators and code in general when e.g., PropertyExpression<Image> could be constructed with literals.

File: include/mbgl/style/expression/image.hpp (hunk 3 above)
Comment (alexshalamov) [reply include/mbgl/style/expression/image.hpp]: renamed toValue

File: include/mbgl/style/expression/type.hpp (hunk 4 above)
Comment (alexshalamov) [reply include/mbgl/style/expression/type.hpp]: so that it can be used as literal in constexpr context
`constexpr ImageType Image;`

File: src/mbgl/renderer/image_manager.cpp (hunk 5 above)
Comment (alexshalamov) [reply src/mbgl/renderer/image_manager.cpp]: done

File: src/mbgl/style/expression/is_constant.cpp (hunk 6 above)
Comment (alexshalamov) [reply src/mbgl/style/expression/is_constant.cpp]: function returns void, setting flag to false is the only way to short-circuit recursion

File: include/mbgl/style/expression/type.hpp (hunk 4 above)
Comment (alexshalamov) [reply include/mbgl/style/expression/type.hpp]: changed to default constructor

File: src/mbgl/style/expression/coercion.cpp [hunk 10]
```diff
@@ -115,6 +120,10 @@ mbgl::Value Coercion::serialize() const {
         serialized.push_back(inputs[0]->serialize());
         serialized.emplace_back(std::unordered_map<std::string, mbgl::Value>());
         return serialized;
+    } else if (getType().is<type::ImageType>()) {
+        std::vector<mbgl::Value> serialized{{std::string("image")}};
```
Comment (pozdnyakov) [inline src/mbgl/style/expression/coercion.cpp]: `std::vector<mbgl::Value> {{std::string("image")}, inputs[0]->serialize() };` ?

File: src/mbgl/style/expression/is_constant.cpp (hunk 6 above)
Comment (pozdnyakov) [reply src/mbgl/style/expression/is_constant.cpp]: :crying_cat_face:

File: src/mbgl/style/layers/symbol_layer_properties.hpp [hunk 11]
```diff
@@ -31,9 +31,9 @@ struct IconIgnorePlacement : LayoutProperty<bool> {
     static bool defaultValue() { return false; }
 };
 
-struct IconImage : DataDrivenLayoutProperty<std::string> {
+struct IconImage : DataDrivenLayoutProperty<expression::Image> {
     static constexpr const char *name() { return "icon-image"; }
-    static std::string defaultValue() { return ""; }
+    static expression::Image defaultValue() { return ""; }
```
Comment (pozdnyakov) [inline src/mbgl/style/layers/symbol_layer_properties.hpp]: nit: `return std::string(); // or {} to obviate conversion from C string to std::string, same in other places`

Comment (alexshalamov) [reply src/mbgl/style/layers/symbol_layer_properties.hpp]: fixed

File: src/mbgl/style/expression/coercion.cpp (hunk 10 above)
Comment (alexshalamov) [reply src/mbgl/style/expression/coercion.cpp]: done

File: test/style/property_expression.test.cpp [hunk 12]
```diff
@@ -30,6 +30,10 @@ static StubGeometryTileFeature oneColor {
     PropertyMap {{ "color", "red"s }}
 };
 
+static StubGeometryTileFeature oneImage{PropertyMap{{"image_name", "maki-11"s}}};
+
+static StubGeometryTileFeature emptyTileFeature{PropertyMap{}};
```
[the remark this answers is not in the document]
Comment (alexshalamov) [reply test/style/property_expression.test.cpp]: :+1:

**Comment by springmeyer** [comment]
> @julianrex @tobrun @springmeyer Could you please take a look at [darwin], [android] and [node] commits.

It looks like the `[node]` changes relate only to the expression API exposed in node and don't impact the node.js rendering logic at all. Is that right? I'm not familiar with the expression API code, so I'm not in a good position to provide a review. Perhaps someone else on @mapbox/gl-core knows what `platform/node/src/node_expression.cpp` is used for? Testing perhaps (before we moved to more C++ tests)?

**Comment by alexshalamov** [comment]
@springmeyer `platform/node/src/node_expression.cpp` was used for expression tests and also can be used to set / get expressions for layer properties at runtime, not sure if anyone is using node expression bindings for runtime styling.
